video/out/hwdec/hwdec_drmprime.c

```diff
@@ -92,6 +92,17 @@ static int init(struct ra_hwdec *hw)
         return -1;
     }
 
+#if HAVE_V4L2REQUEST
+    /*
+     * AVCodecHWConfig contains a combo of a pixel format and hwdevice type,
+     * correct type must be created here or hwaccel will fail.
+     *
+     * FIXME: Create hwdevice based on type in AVCodecHWConfig
+     */
+    int ret = av_hwdevice_ctx_create(&p->hwctx.av_device_ref,
+                                     AV_HWDEVICE_TYPE_V4L2REQUEST,
+                                     NULL, NULL, 0);
+#else
```

Comment:
I know it's a draft but the main thing is that if it really ends up being a completely distinct hwdevice type, we will want a separate hwdec for it. Of course, you can refactor to share code but this #ifdef approach cannot actually be used because, as you say, it breaks `hwdec_drmprime` for everything else.

Reply:
Agree, and this is what really confused me when testing the v4l2reqeust hwdevice changes with mpv, this file seem more related to the `imgfmt` and not the `hwcontext` type when using a `--hwdec=v4l2request` parameter.

Testing this with or without the changes in this file you must still use `--hwdec=v4l2request` to use the new `hwcontext`, the only difference was if the hwaccel in ffmpeg would work or if it would fail and complain that an unexpected hwdevice type is used.

On e.g. a RPi5 board I can use following to playback a HEVC video file (after #14508).
```
mpv --hwdec=v4l2request --hwdec-codecs=all --gpu-hwdec-interop=drmprime-overlay --drm-draw-plane=overlay --drm-drmprime-video-plane=primary <hevc file>
```

Reply:
> Still, you'll want to have a distinct hwdec and refactor the code to be shared. We should not fudge the relationship between hwdecs and hwaccels.

I think this is where there may be an architectural flaw in mpv, the hwdec parameter seem to be more tied to what hwcontext gets used not what hwdec is used, the hwdec used seem more tied to the imgfmt of the hwcontext.

There is also a `create_dev` ops on the hwcontext that probably could/should be used at hwdec level instead of duplicated/separate logic. Ideally the hwcontext could/should be passed along as parameter/input to the hwdec. If mpv wants to treat the hwdec as the hwcontext+imgfmt combo that is also fine.

I am really hoping that a mpv developer will eventually continue working on any changes required based on the outcome of ffmpeg reviews, mainly wanted to start this discussion early so that others can start preparing for upcoming changes, and I was initially expecting that mpv already had a separation between hwdevice and hwdec/imgfmt :smile:

The code here is more minimum effort to make the new hwdevice type work and limit the bad impact to when mpv is built with `--enable-v4l2request`.